p2p/stream.go

```diff
@@ -0,0 +1,88 @@
+package p2p
+
+import (
+	"io"
+	"sync"
+
+	manet "gx/ipfs/QmNqRnejxJxjRroz7buhrjfU8i3yNBLa81hFtmf2pXEffN/go-multiaddr-net"
+	ma "gx/ipfs/QmUxSEGbv2nmYNnfXi7839wwQqTN3kwQeUxe8dTjZWZs7J/go-multiaddr"
+	net "gx/ipfs/QmXdgNhVEgjLxjUoMs5ViQL7pboAt3Y7V7eGHRiE4qrmTE/go-libp2p-net"
+	"gx/ipfs/QmZNkThpqfVXs9GNbexPrfBbXSLNYeKrE7jwFM2oqHbyqN/go-libp2p-protocol"
+)
+
+// Stream holds information on active incoming and outgoing p2p streams.
+type Stream struct {
+	id uint64
+
+	Protocol protocol.ID
+
+	OriginAddr ma.Multiaddr
+	TargetAddr ma.Multiaddr
+
+	Local  manet.Conn
+	Remote net.Stream
+
+	Registry *StreamRegistry
+}
+
+// Close closes stream endpoints and deregisters it
+func (s *Stream) Close() error {
+	s.Local.Close()
+	s.Remote.Close()
+	s.Registry.Deregister(s.id)
+	return nil
+}
+
+// Reset closes stream endpoints and deregisters it
+func (s *Stream) Reset() error {
+	s.Local.Close()
+	s.Remote.Reset()
+	s.Registry.Deregister(s.id)
+	return nil
+}
+
+func (s *Stream) startStreaming() {
+	go func() {
+		_, err := io.Copy(s.Local, s.Remote)
+		if err != nil {
+			s.Reset()
+		} else {
+			s.Close()
```

Comment:
Ok. *Now* I remember the issue here. We can't always call `Close()` on a stream from one thread while another is writing (some transports will have memory races if we do that).

Regardless, given that we can't really "nicely" shut down the connection, maybe we should just reset the stream? This is really bugging me. TCP supports the following usecase:

1. Client: Write and finish with an EOF.
2. Server: Read until EOF, handle request, write and finish with an EOF.
3. Client: Read response until EOF.

As it stands, these streams won't support this.

---

So, for now, maybe we should go back to simply resetting the stream and documenting the limitation. In the future, we can try exposing this half-closed feature for transports that support it.

(we should also put a big fat comment here explaining this)

Reply:
> Regardless, given that we can't really "nicely" shut down the connection, maybe we should just reset the stream?

Just doing resets here breaks the tests in some quite hard to debug ways. I'd leave it this way for now and come back once the other half-open stuff is done

Reply:
Sounds good. Basically, as it stands, the "correct" way to use this feature is to negotiate a connection finish at the application layer before closing the connection. With the multiaddr-net changes, that shouldn't be necessary.